### PR TITLE
TAJO-1436 Add Bind method to EvalNode

### PR DESCRIPTION
**Pull request by navis**
eval(Schema schema, Tuple tuple) --> bind(Schema schema) + eval(Tuple tuple)

### CODE REVIEW

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/HashLeftOuterJoinExec.java [hunk 1]
```diff
@@ -120,11 +120,11 @@ public HashLeftOuterJoinExec(TaskAttemptContext context, JoinNode plan, Physical
     leftKeyTuple = new VTuple(leftKeyList.length);
 
     rightNumCols = rightChild.getSchema().size();
-  }
 
-  @Override
-  protected void compile() {
-    joinQual = context.getPrecompiledEval(inSchema, joinQual);
```
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/HashLeftOuterJoinExec.java]: The compile() function should not be removed.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java [hunk 2]
```diff
@@ -69,9 +70,8 @@ public SeqScanExec(TaskAttemptContext context, ScanNode plan,
     this.qual = plan.getQual();
     this.fragments = fragments;
 
-    if (fragments != null
-        && plan.getTableDesc().hasPartition()
```
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java]: The variable ```fragments``` can be null, and ```rewriteColumnPartitionedTableSchema()``` should not be called in that case.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/HashLeftOuterJoinExec.java (hunk 1 above)
Comment (navis) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/HashLeftOuterJoinExec.java]: Yes, mistake. Nice catch.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java [hunk 3]
```diff
@@ -197,6 +206,9 @@ public Tuple next() throws IOException {
     if (fragments == null) {
       return null;
     }
+    if (scanner == null) {
+      System.out.println("[SeqScanExec/next] ");
```
Comment (jihoonson) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java]: Please remove debug codes.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java (hunk 2 above)
Comment (navis) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java]: Even fragments is null, partition columns in expr should be replaced with null. Without this, some tests fails with exception("No Such Column Reference").

File: tajo-plan/src/main/java/org/apache/tajo/plan/expr/FunctionEval.java [hunk 4]
```diff
@@ -40,13 +40,27 @@
   @Expose protected FunctionDesc funcDesc;
 	@Expose protected EvalNode [] argEvals;
 
+  private transient Tuple params;
```
Comment (jihoonson) [inline tajo-plan/src/main/java/org/apache/tajo/plan/expr/FunctionEval.java]: This variable is missed in ```hashCode()```, ```clone()```, and ```equals()``` functions.

Comment (navis) [inline tajo-plan/src/main/java/org/apache/tajo/plan/expr/FunctionEval.java]: It's transient field as it declared, which should not be included in hash or equals calculation.

File: tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java (hunk 3 above)
Comment (navis) [inline tajo-core/src/main/java/org/apache/tajo/engine/planner/physical/SeqScanExec.java]: done

File: tajo-plan/src/main/java/org/apache/tajo/plan/expr/FunctionEval.java (hunk 4 above)
Comment (jihoonson) [inline tajo-plan/src/main/java/org/apache/tajo/plan/expr/FunctionEval.java]: Oops. Sorry.

**Comment by jihoonson** [comment]
Thanks @navis. The patch looks good.
I have some additional comments as follows.
* Changes on NettyClientBase look for https://issues.apache.org/jira/browse/TAJO-1424. If so, it would be better to resolve in that issue. Please remove them.
* The bind() function must be called before eval(). So, it would be much great if you add a routine to check the binding state.

**Comment by navis** [comment]
I've also thought about that (bind check), but thinking that the intention of this issue was for removing any non-working code from eval, I decided to remove it. Opinion?

I'll fix TAJO-1424, which seemed abondoned for a long time.

**Comment by jihoonson** [comment]
Ok. I created another issue for bind check. https://issues.apache.org/jira/browse/TAJO-1499

**Comment by jihoonson** [comment]
Would you please rebase?

**Comment by jihoonson** [comment]
Oh, you have already rebased. Sorry. I'll review ASAP.

**Pull request closed without merging**